src/cljam/io/util/bin.clj

```diff
@@ -5,39 +5,59 @@
 
 (defn- reg->bins*
   "Returns candidate bins for the specified region as a vector."
-  [^long beg ^long end]
-  (let [max-pos 0x1FFFFFFF
-        beg (if (<= beg 0) 0 (bit-and (dec beg) max-pos))
-        end (if (<= end 0) max-pos (bit-and (dec end) max-pos))]
-    (if (<= beg end)
-      (loop [bins (transient [0])
-             xs [[1 26] [9 23] [73 20] [585 17] [4681 14]]]
-        (if-let [[^long ini shift] (first xs)]
-          (let [ini* (+ ini (bit-shift-right beg shift))
-                end* (+ ini (bit-shift-right end shift))]
-            (recur
-             (loop [b bins k ini*]
-               (if (<= k end*)
-                 (recur (conj! b k) (inc k))
-                 b))
-             (next xs)))
-          (persistent! bins))))))
+  [^long beg ^long end ^long min-shift ^long depth]
+  (let [max-pos (bit-shift-left 1 (+ min-shift (* 3 depth)))
+        beg (if (<= beg 0) 0 (min (dec beg) max-pos))
+        end (if (<= end 0) max-pos (min end max-pos))]
+    (into [0]
+          (mapcat
+           (fn [^long d]
+             (let [t
+                   (apply + (map (fn [^long x] (bit-shift-left 1 (* x 3)))
+                                 (range (inc d))))
+                   s (+ min-shift (* 3 (- depth d 1)))]
+               (range (+ t (bit-shift-right beg s))
+                      (+ t 1 (bit-shift-right (dec end) s))))))
+          (range depth))))
+
+(defn- first-bin-of-level ^long [^long level]
+  (if (zero? level)
+    0
+    (unsigned-bit-shift-right
+         ;; 2r1001001001001001001001001001001001001001001001001001001001001001
+     -0x6db6db6db6db6db7
+     (- 64 (inc (* (dec level) 3))))))
+
+(defn- bin-width-of-level ^long [^long level ^long min-shift ^long depth]
+  (bit-shift-left 1 (+ min-shift (* 3 (- depth level)))))
+
+(defn- bin-level ^long [^long bin]
+  (let [x (inc (quot (- 64 (Long/numberOfLeadingZeros bin)) 3))]
+    (cond-> x
+      (< bin (first-bin-of-level x)) dec)))
+
+(defn bin-beg ^long [^long bin ^long min-shift ^long depth]
+  (let [level (bin-level bin)]
+    (inc (* (- bin (first-bin-of-level level))
+            (bin-width-of-level level min-shift depth)))))
 
 (def ^:private reg->bins (memoize reg->bins*))
```

Comment:
The position of defining memoized version is a little too far from original version.